File: app/src/main/java/dev/ragnarok/fenrir/util/AppTextUtils.java
package dev.ragnarok.fenrir.util;

import android.content.Context;
import android.text.TextUtils;

import androidx.annotation.Nullable;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.regex.Pattern;

import dev.ragnarok.fenrir.R;

public class AppTextUtils {
    public static final long ONE_DAY_SEC = 24 * 60 * 60;
    private static final Date DATE = new Date();
    private static final Calendar CALENDAR = Calendar.getInstance();
    private static final String K = "K";
    private static final String ZERO = "0";
    private static final String TWO_ZERO = "00";
    private static final String POINT = ".";
    private static final String EMPTY = "";
    private static SimpleDateFormat SHORT_DATE = new SimpleDateFormat("HH:mm", Utils.getAppLocale());
    private static SimpleDateFormat FULL_DATE = new SimpleDateFormat("dd MMM yyyy HH:mm", Utils.getAppLocale());
    private static SimpleDateFormat FULL_LITTLE_DATE = new SimpleDateFormat("dd.MM.yy HH:mm", Utils.getAppLocale());

    public static void updateDateLang(Locale locale) {
        SHORT_DATE = new SimpleDateFormat("HH:mm", locale);
        FULL_DATE = new SimpleDateFormat("dd MMM yyyy HH:mm", locale);
        FULL_LITTLE_DATE = new SimpleDateFormat("dd.MM.yy HH:mm", locale);
    }

    public static String safeTrim(String text, @Nullable String ifNull) {
        return text == null ? ifNull : text.trim();
    }

    public static String reduceStringForPost(String input) {
        return reduceText(input, 400);
    }

    public static String reduceText(String text, int maxLenght) {
        if (text == null || text.length() < maxLenght) {
            return text;
        } else {
            return text.substring(0, maxLenght).concat("...");
        }
    }

    /**
     * Получения строки с размером в мегабайтах
     *
     * @param size размер в байтах
     * @return строка типа "13.6 Mb"
     */
    public static String getSizeString(long size) {
        double sizeDouble = ((double) size) / 1024 / 1024;
        double newDouble = new BigDecimal(sizeDouble).setScale(2, RoundingMode.UP).doubleValue();
        return newDouble + " Mb";
    }

    /**
     * Получение строки с датой и временем сообщений
     *
     * @param unixTime дата в формате unix-time
     * @return строка с датой и временем
     */
    public static String getDateFromUnixTime(long unixTime) {
        CALENDAR.setTimeInMillis(System.currentTimeMillis());

        DATE.setTime(unixTime * 1000);
        CALENDAR.set(CALENDAR.get(Calendar.YEAR), CALENDAR.get(Calendar.MONTH), CALENDAR.get(Calendar.DATE), 0, 0, 0);
        return unixTime * 1000 > CALENDAR.getTimeInMillis() ? SHORT_DATE.format(DATE) : FULL_DATE.format(DATE);
    }

    /**
     * Получение строки с датой и временем сообщений
     *
     * @param unixTime дата в формате unix-time
     * @return строка с датой и временем
     */
    public static String getDateFromUnixTime(Context context, long unixTime) {
        Calendar calendar = Calendar.getInstance();
        calendar.set(calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DATE), 0, 0, 0);

        DATE.setTime(unixTime * 1000);

        long startToday = calendar.getTimeInMillis() / 1000;
        long startTomorrow = startToday + ONE_DAY_SEC;
        long startOfDayAfterTomorrow = startTomorrow + ONE_DAY_SEC;
        long startOfYesterday = startToday - ONE_DAY_SEC;

        if (unixTime >= startToday && unixTime < startTomorrow) {
            return context.getString(R.string.formatted_date_today, SHORT_DATE.format(DATE));
        }

        if (unixTime >= startOfYesterday && unixTime < startToday) {
            return context.getString(R.string.formatted_date_yesterday, SHORT_DATE.format(DATE));
        }

        if (unixTime >= startTomorrow && unixTime < startOfDayAfterTomorrow) {
            return context.getString(R.string.formatted_date_tomorrow, SHORT_DATE.format(DATE));
        }

        return FULL_DATE.format(DATE);
    }

    /**
     * Получение строки с датой и временем сообщений
     *
     * @param unixTime дата в формате unix-time
     * @return строка с датой и временем
     */
    public static String getDateFromUnixTimeShorted(Context context, long unixTime) {
        Calendar calendar = Calendar.getInstance();
        calendar.set(calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DATE), 0, 0, 0);

        DATE.setTime(unixTime * 1000);

        long startToday = calendar.getTimeInMillis() / 1000;
        long startTomorrow = startToday + ONE_DAY_SEC;
        long startOfDayAfterTomorrow = startTomorrow + ONE_DAY_SEC;
        long startOfYesterday = startToday - ONE_DAY_SEC;

        if (unixTime >= startToday && unixTime < startTomorrow) {
            return SHORT_DATE.format(DATE);
        }

        if (unixTime >= startOfYesterday && unixTime < startToday) {
            return context.getString(R.string.formatted_date_yesterday_clean, SHORT_DATE.format(DATE));
        }

        if (unixTime >= startTomorrow && unixTime < startOfDayAfterTomorrow) {
            return context.getString(R.string.formatted_date_tomorrow_clean, SHORT_DATE.format(DATE));
        }

        return FULL_LITTLE_DATE.format(DATE);
    }

    public static String getDurationString(int seconds) {
        int hours = seconds / 3600;
        int minutes = (seconds % 3600) / 60;
        seconds = seconds % 60;
        if (hours == 0) {
            return twoDigitString(minutes) + ":" + twoDigitString(seconds);
        } else {
            return twoDigitString(hours) + ":" + twoDigitString(minutes) + ":" + twoDigitString(seconds);
        }
    }

    public static String getDurationStringMS(int ms) {
        return getDurationString(ms / 1000);
    }

    private static String twoDigitString(int number) {
        if (number == 0) {
            return TWO_ZERO;
        }

        if (number / 10 == 0) {
            return ZERO + number;
        }

        return String.valueOf(number);
    }

    public static String getCounterWithK(int counter) {
        int num = counter / 1000;
        if (num >= 1 && num < 10) {
            //return num + "." + (counter / 100) % 10 + K;
            return String.valueOf(counter);
        }
        if (num >= 10) {
            return num + K;
        }

        return String.valueOf(counter);
    }

    public static String getDateWithZeros(String date) {
        if (TextUtils.isEmpty(date) || date.equals(ZERO)) {
            return EMPTY;
        }

        String[] tmp = date.split(Pattern.quote(POINT));
        for (int i = 0; i < tmp.length; i++) {
            if (tmp[i].length() == 1) {
                tmp[i] = ZERO + tmp[i];
            }
        }

        return TextUtils.join(POINT, tmp);
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/util/AppTextUtils.kt
package dev.ragnarok.fenrir.util

import android.content.Context
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.link.internal.OwnerLinkSpanFactory
import java.math.BigDecimal
import java.math.RoundingMode
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.regex.Pattern

object AppTextUtils {
    private const val ONE_DAY_SEC = (24 * 60 * 60).toLong()
    private val DATE = Date()
    private val CALENDAR = Calendar.getInstance()
    private const val K = "K"
    private const val ZERO = "0"
    private const val TWO_ZERO = "00"
    private const val POINT = "."
    private const val EMPTY = ""
    private var SHORT_DATE = SimpleDateFormat("HH:mm", Utils.appLocale)
    private var FULL_DATE = SimpleDateFormat("dd MMM yyyy HH:mm", Utils.appLocale)
    private var FULL_LITTLE_DATE = SimpleDateFormat("dd.MM.yy HH:mm", Utils.appLocale)


    fun updateDateLang(locale: Locale?) {
        SHORT_DATE = SimpleDateFormat("HH:mm", locale)
        FULL_DATE = SimpleDateFormat("dd MMM yyyy HH:mm", locale)
        FULL_LITTLE_DATE = SimpleDateFormat("dd.MM.yy HH:mm", locale)
    }


    fun safeTrim(text: String?, ifNull: String?): String? {
        return text?.trim { it <= ' ' } ?: ifNull
    }


    fun reduceStringForPost(input: String?): String? {
        val pp = OwnerLinkSpanFactory.findPatterns(input, owners = true, topics = false)
        var l = 400
        for (i in pp.orEmpty()) {
            if (i.start >= l) {
                break
            } else if (i.end > l) {
                l = i.end + 1
                break
            }

        }
        return reduceText(input, l)
    }


    fun reduceText(text: String?, maxLength: Int): String? {
        return if (text == null || text.length < maxLength) {
            text
        } else {
            text.substring(0, maxLength) + "..."
        }
    }

    /**
     * Получения строки с размером в мегабайтах
     *
     * @param size размер в байтах
     * @return строка типа "13.6 Mb"
     */

    fun getSizeString(size: Long): String {
        val sizeDouble = size.toDouble() / 1024 / 1024
        val newDouble = BigDecimal(sizeDouble).setScale(2, RoundingMode.UP).toDouble()
        return "$newDouble Mb"
    }

    /**
     * Получение строки с датой и временем сообщений
     *
     * @param unixTime дата в формате unix-time
     * @return строка с датой и временем
     */

    fun getDateFromUnixTime(unixTime: Long): String {
        CALENDAR.timeInMillis = System.currentTimeMillis()
        DATE.time = unixTime * 1000
        CALENDAR[CALENDAR[Calendar.YEAR], CALENDAR[Calendar.MONTH], CALENDAR[Calendar.DATE], 0, 0] =
            0
        return if (unixTime * 1000 > CALENDAR.timeInMillis) SHORT_DATE.format(DATE) else FULL_DATE.format(
            DATE
        )
    }

    /**
     * Получение строки с датой и временем сообщений
     *
     * @param unixTime дата в формате unix-time
     * @return строка с датой и временем
     */

    fun getDateFromUnixTime(context: Context, unixTime: Long): String {
        val calendar = Calendar.getInstance()
        calendar[calendar[Calendar.YEAR], calendar[Calendar.MONTH], calendar[Calendar.DATE], 0, 0] =
            0
        DATE.time = unixTime * 1000
        val startToday = calendar.timeInMillis / 1000
        val startTomorrow = startToday + ONE_DAY_SEC
        val startOfDayAfterTomorrow = startTomorrow + ONE_DAY_SEC
        val startOfYesterday = startToday - ONE_DAY_SEC
        if (unixTime in startToday until startTomorrow) {
            return context.getString(R.string.formatted_date_today, SHORT_DATE.format(DATE))
        }
        if (unixTime in startOfYesterday until startToday) {
            return context.getString(R.string.formatted_date_yesterday, SHORT_DATE.format(DATE))
        }
        return if (unixTime in startTomorrow until startOfDayAfterTomorrow) {
            context.getString(
                R.string.formatted_date_tomorrow,
                SHORT_DATE.format(DATE)
            )
        } else FULL_DATE.format(DATE)
    }

    /**
     * Получение строки с датой и временем сообщений
     *
     * @param unixTime дата в формате unix-time
     * @return строка с датой и временем
     */

    fun getDateFromUnixTimeShorted(context: Context, unixTime: Long): String {
        val calendar = Calendar.getInstance()
        calendar[calendar[Calendar.YEAR], calendar[Calendar.MONTH], calendar[Calendar.DATE], 0, 0] =
            0
        DATE.time = unixTime * 1000
        val startToday = calendar.timeInMillis / 1000
        val startTomorrow = startToday + ONE_DAY_SEC
        val startOfDayAfterTomorrow = startTomorrow + ONE_DAY_SEC
        val startOfYesterday = startToday - ONE_DAY_SEC
        if (unixTime in startToday until startTomorrow) {
            return SHORT_DATE.format(DATE)
        }
        if (unixTime in startOfYesterday until startToday) {
            return context.getString(
                R.string.formatted_date_yesterday_clean,
                SHORT_DATE.format(DATE)
            )
        }
        return if (unixTime in startTomorrow until startOfDayAfterTomorrow) {
            context.getString(
                R.string.formatted_date_tomorrow_clean,
                SHORT_DATE.format(DATE)
            )
        } else FULL_LITTLE_DATE.format(DATE)
    }

    fun getDurationString(seconds: Int): String {
        var pSeconds = seconds
        val hours = pSeconds / 3600
        val minutes = pSeconds % 3600 / 60
        pSeconds %= 60
        return if (hours == 0) {
            twoDigitString(minutes) + ":" + twoDigitString(pSeconds)
        } else {
            twoDigitString(hours) + ":" + twoDigitString(minutes) + ":" + twoDigitString(
                pSeconds
            )
        }
    }


    fun getDurationStringMS(ms: Int): String {
        return getDurationString(ms / 1000)
    }

    private fun twoDigitString(number: Int): String {
        if (number == 0) {
            return TWO_ZERO
        }
        return if (number / 10 == 0) {
            ZERO + number
        } else number.toString()
    }


    fun getCounterWithK(counter: Int): String {
        val num = counter / 1000
        if (num in 1..9) {
            //return num + "." + (counter / 100) % 10 + K;
            return counter.toString()
        }
        return if (num >= 10) {
            num.toString() + K
        } else counter.toString()
    }


    fun getDateWithZeros(date: String?): String {
        if (date.isNullOrEmpty() || date == ZERO) {
            return EMPTY
        }
        val tmp = date.split(Pattern.quote(POINT)).toTypedArray()
        for (i in tmp.indices) {
            if (tmp[i].length == 1) {
                tmp[i] = ZERO + tmp[i]
            }
        }
        return Utils.join(POINT, tmp)
    }
}